File: 23.合并k个升序链表.java
/*
 * @lc app=leetcode.cn id=23 lang=java
 *
 * [23] 合并K个升序链表
 */
import java.util.*;
// @lc code=start
/**
 * Definition for singly-linked list.
 * 
 */
class Solution {
    // siftUp in Priority Queue
    // private static <T> void siftUpUsingComparator(
    //     int k, T x, Object[] es, Comparator<? super T> cmp) {
    //     while (k > 0) {
    //         int parent = (k - 1) >>> 1;
    //         Object e = es[parent];
    //         if (cmp.compare(x, (T) e) >= 0)
    //             break;
    //         es[k] = e;
    //         k = parent;
    //     }
    //     es[k] = x;
    // }

    
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0)
            return null;
        ListNode dummy = new ListNode(-1), p = dummy;
        PriorityQueue<ListNode> pq = new PriorityQueue<>
        (lists.length, (ListNode a, ListNode b) -> (a.val - b.val));
        for (ListNode head : lists) {
            if (head != null)
                pq.add(head);
        }
        while (!pq.isEmpty()) {
            ListNode node = pq.poll();
            p.next = node;
            p = p.next;
            if (node.next != null)
                pq.add(node.next);
        }
        return dummy.next;
    }
}
// @lc code=end

File: 23.合并k个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并K个升序链表
 */

#include<string>
#include<algorithm>
#include<vector>
#include<queue>
using namespace std;
// @lc code=start
/**
 * Definition for singly-linked list.
 * 
 * */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    struct greater_ListNode {
        ListNode *node;
        greater_ListNode(ListNode *nd) : node(nd) {}
        bool operator< (const greater_ListNode &another) const {
            return node->val > another.node->val;
        }
    };

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode *dummy = new ListNode(-1), *p = dummy;
        priority_queue<greater_ListNode> pq;
        for (auto listnode : lists)
            if (listnode != nullptr)
                pq.emplace(greater_ListNode(listnode));
        while (!pq.empty()) {
            ListNode *cur = pq.top().node;
            pq.pop();
            p->next = cur;
            p = p->next;
            if (cur->next != nullptr) {
                pq.emplace(greater_ListNode(cur->next));
            }
        }
        return dummy->next;
    }
};
// @lc code=end

